test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs

```diff
@@ -0,0 +1,146 @@
+            DateTime edgeHubRestartedTime,
+            HttpStatusCode edgeHubRestartStatusCode,
+            uint restartSequenceNumber,
+            ILogger logger)
+        {
+            this.DmModuleClient = Preconditions.CheckNotNull(dmModuleClient, nameof(dmModuleClient));
+            this.BatchId = batchId;
+            this.RunExpirationTime = runExpirationTime;
+            this.CancellationToken = Preconditions.CheckNotNull(cancellationToken, nameof(cancellationToken));
+            this.EdgeHubRestartedTime = edgeHubRestartedTime;
+            this.EdgeHubRestartStatusCode = edgeHubRestartStatusCode;
+            this.RestartSequenceNumber = Preconditions.CheckNotNull(restartSequenceNumber, nameof(restartSequenceNumber));
+            this.Logger = Preconditions.CheckNotNull(logger, nameof(logger));
+        }
+
+        public async Task ConnectAsync()
+        {
+            (DateTime dmCompletedTime, HttpStatusCode dmStatusCode) = await this.SendDirectMethodAsync(
+                Settings.Current.DeviceId,
+                Settings.Current.DirectMethodTargetModuleId,
+                this.DmModuleClient,
+                Settings.Current.DirectMethodName,
+                this.RunExpirationTime,
+                this.CancellationToken,
+                this.Logger).ConfigureAwait(false);
+
+            TestResultBase dmTestResult = new EdgeHubRestartDirectMethodResult(
+                Settings.Current.ModuleId + "." + TestOperationResultType.EdgeHubRestartDirectMethod.ToString(),
+                DateTime.UtcNow,
+                Settings.Current.TrackingId,
+                this.BatchId,
+                Interlocked.Read(ref this.directMethodCount).ToString(),
+                this.EdgeHubRestartedTime,
+                this.EdgeHubRestartStatusCode,
+                dmCompletedTime,
+                dmStatusCode,
+                this.RestartSequenceNumber);
+
+            var reportClient = new TestResultReportingClient { BaseUrl = Settings.Current.ReportingEndpointUrl.AbsoluteUri };
+            await ModuleUtil.ReportTestResultUntilSuccessAsync(
+                reportClient,
+                this.Logger,
+                dmTestResult,
+                this.CancellationToken).ConfigureAwait(false);
+        }
+
+        async Task<Tuple<DateTime, HttpStatusCode>> SendDirectMethodAsync(
+            string deviceId,
+            string targetModuleId,
+            ModuleClient moduleClient,
+            string directMethodName,
+            DateTime runExpirationTime,
+            CancellationToken cancellationToken,
+            ILogger logger)
+        {
+            while ((!cancellationToken.IsCancellationRequested) && (DateTime.UtcNow < runExpirationTime))
+            {
+                try
+                {
+                    // Direct Method sequence number is always increasing regardless of sending result.
+                    Interlocked.Increment(ref this.directMethodCount);
```

Comment:
why do you increment sequence number all the time during retry?

Reply:
why not? This make sure the seqeunce number is unique from point of view of both sender and receiver.

Reply:
This is applicable for DM case. Added another comment in Messaging case to NOT increment the seqeunce number if failure.